File: src/main/scala/com/formious/models/AWSAccountAdministrator.scala
package com.formious.models

import java.time.ZonedDateTime
import java.sql.ResultSet
import com.formious.common.Database.query
import com.formious.common.Recoders._

case class AWSAccountAdministrator(id: Int,
                                   aws_account_id: Int,
                                   administrator_id: Int,
                                   priority: Int,
                                   created: ZonedDateTime)

object AWSAccountAdministrator {
  def apply(row: ResultSet) = new AWSAccountAdministrator(
    row.getInt("id"),
    row.getInt("aws_account_id"),
    row.getInt("administrator_id"),
    row.getInt("priority"),
    row.getTimestamp("created").toZonedDateTime)

  def all = {
    query("SELECT * FROM aws_account_administrator")(AWSAccountAdministrator.apply)
  }

  def create(aws_account_id: Int, administrator_id: Int, priority: Option[Int]) = {
    query("""
      INSERT INTO aws_account_administrator (aws_account_id, administrator_id, priority)
      VALUES (?, ?, ?) RETURNING *
    """, List(aws_account_id, administrator_id, priority))(AWSAccountAdministrator.apply).head
  }

  def delete(aws_account_id: Int, administrator_id: Int) = {
    query("""
      DELETE FROM aws_account_administrator
      WHERE aws_account_id = ? AND administrator_id = ?
    """, List(aws_account_id, administrator_id))(AWSAccountAdministrator.apply).head
  }
}

File: src/com/formious/db/aws_account_administrator.clj
(ns com.formious.db.aws-account-administrator
  (:require [com.formious.db :as db])
  (:import [java.time ZonedDateTime]))

; Int Int Int Int ZonedDateTime
(defrecord AWSAccountAdministrator [id aws_account_id administrator_id priority created])

(defn row->AWSAccountAdministrator
  [row]
  (map->AWSAccountAdministrator (update row :created .toZonedDateTime)))

(defn all
  []
  (->> (db/query "SELECT * FROM aws_account_administrator")
       (map row->AWSAccountAdministrator)))

(defn insert!
  ; (aws_account_id: Int, administrator_id: Int, priority: Option[Int])
  [row]
  (->> row (db/insert! "aws_account_administrator") row->AWSAccountAdministrator))

(defn delete!
  [aws_account_id administrator_id]
  (db/delete! "aws_account_administrator" ["aws_account_id = ? AND administrator_id = ?" aws_account_id, administrator_id]))
